component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java

```diff
@@ -0,0 +1,249 @@
+            int digit = (int) (num % 10);
+            num /= 10;
+            if (doubleDigit) {
+                digit = CHECKSUM_DIGITS[digit];
+            }
+            total += digit;
+            doubleDigit = !doubleDigit;
+        }
+        int result = total % 10;
+        if (result > 0) {
+            result = 10 - result;
+        }
+        return result;
+    }
+
+    /**
+     * This method uses the JCE to provide the HMAC-SHA-1
+     * algorithm. HMAC computes a Hashed Message Authentication Code and in this
+     * case SHA1 is the hash algorithm used.
+     *
+     * @param keyBytes Bytes to use for the HMAC-SHA-1 key.
+     * @param text     Message or text to be authenticated.
+     * @throws NoSuchAlgorithmException If no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
+     */
+    public static byte[] hmacShaGenerate(byte[] keyBytes, byte[] text) throws NoSuchAlgorithmException,
+            InvalidKeyException {
+
+        Mac hmacSha;
+        try {
+            hmacSha = Mac.getInstance(Constants.ALGORITHM_HMAC);
+        } catch (NoSuchAlgorithmException nsa) {
+            hmacSha = Mac.getInstance(Constants.ALGORITHM_HMAC_SHA);
+        }
+        SecretKeySpec macKey = new SecretKeySpec(keyBytes, "RAW");
+        hmacSha.init(macKey);
+        return hmacSha.doFinal(text);
+    }
+
+    /**
+     * This method generates an OTP value for the given set of parameters.
+     *
+     * @param secret           The shared secret.
+     * @param movingFactor     The counter, or other value that changes on a per-use
+     *                         basis.
+     * @param codeDigits       The number of digits in the OTP, not including the checksum,
+     *                         if any.
+     * @param addChecksum      A flag that indicates if a checksum digit
+     *                         should be appended to the OTP.
+     * @param truncationOffset The offset into the MAC result to begin truncation. If this
+     *                         value is out of the range of 0 ... 15, then dynamic truncation
+     *                         will be used. Dynamic truncation is when the last 4 bits of
+     *                         the last byte of the MAC are used to determine the start
+     *                         offset.
+     * @throws NoSuchAlgorithmException If no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
+     */
+    public static String generateOTP(byte[] secret, long movingFactor, int codeDigits, boolean addChecksum,
+                                     int truncationOffset) throws NoSuchAlgorithmException, InvalidKeyException {
+
+        // put movingFactor value into text byte array
```

Comment:
Start comments in capital letters and stop by fullstop 
```suggestion
        // Put movingFactor value into text byte array.
```

Reply:
Fix other comments as well

Reply:
fixed in 0e37ffc